app-backend/api/src/main/scala/uploads/Routes.scala

```diff
@@ -51,240 +56,300 @@ trait UploadRoutes
+              newUpload.uploadType,
+              newUpload.source,
+              newUpload.fileType
+            ) match {
+              case (UploadType.S3, _, FileType.NonSpatial) => {
+                if (newUpload.files.nonEmpty) newUpload
                 else
                   throw new IllegalStateException(
-                    "No acceptable files found in the provided source"
+                    "S3 upload must specify a source if no files are specified"
                   )
               }
-            }
-            case (UploadType.S3, None, _) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else
-                throw new IllegalStateException(
-                  "S3 upload must specify a source if no files are specified"
-                )
-            }
-            case (_, _, FileType.GeoJson) => {
-              throw new IllegalStateException(
-                "Scene uploads must contain imagery, not GeoJson"
-              )
-            }
-            case (_, _, _) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else
+              case (UploadType.S3, Some(source), _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else {
+                  val files = listAllowedFilesInS3Source(source)
+                  if (files.nonEmpty) newUpload.copy(files = files)
+                  else
+                    throw new IllegalStateException(
+                      "No acceptable files found in the provided source"
+                    )
+                }
+              }
+              case (UploadType.S3, None, _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else
+                  throw new IllegalStateException(
+                    "S3 upload must specify a source if no files are specified"
+                  )
+              }
+              case (_, _, FileType.GeoJson) => {
                 throw new IllegalStateException(
-                  "Remote repository upload must specify some ids or files"
+                  "Scene uploads must contain imagery, not GeoJson"
                 )
+              }
+              case (_, _, _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else
+                  throw new IllegalStateException(
+                    "Remote repository upload must specify some ids or files"
+                  )
+              }
             }
-          }
 
-        val bytesUploaded = Upload.getBytesUploaded(
-          s3.client,
-          dataBucket,
-          newUpload.files,
-          newUpload.uploadStatus,
-          newUpload.uploadType
-        )
+          onComplete {
+            for {
+              isBelow <- isBelowLimitCheck(
+                userBytesUploaded map { bytes =>
+                  bytes + potentialNewBytes
+                },
+                Domain.Uploads,
+                Action.Create,
+                user
```

Comment:
You can do this with another authorization directive instead of hand-rolling the rejection. You could call `authorizeScopeLimit` after the entity has been decoded and return the `Future` holding whether the upload should proceed. Then the route logic could just be the happy path and `isBelowLimitCheck` could be private / not escape `CommonHandlers`. I think that would be good, because it would avoid mixing authorization and happy path response logic.